domain/network/state/linklayer_merge.go

```diff
@@ -800,3 +848,143 @@ func (st *State) validateSubnetUpdate(ctx context.Context, tx *sqlair.TX, addres
 	}
 	return candidateSubnet.ID.String(), nil
 }
+
+// subnetCIDRUUIDByProviderID returns a map of subnet provider IDs to a
+// struct including the subnet CIDR and UUID.
+func (st *State) subnetCIDRUUIDByProviderID(
+	ctx context.Context,
+	tx *sqlair.TX,
+	add map[string][]mergeAddress,
+) (map[string]providerSubnetCIDR, error) {
+	type ids []string
+	input := make(ids, 0)
+	for _, addrs := range add {
+		for _, addr := range addrs {
+			input = append(input, addr.ProviderSubnetID)
+		}
+	}
+	stmt, err := st.Prepare(`
+SELECT &providerSubnetCIDR.*
+FROM provider_subnet AS ps
+JOIN subnet AS s ON ps.subnet_uuid = s.uuid
+WHERE ps.provider_id IN ($ids[:])
+`, providerSubnetCIDR{}, ids{})
+	if err != nil {
+		return nil, errors.Errorf("preparing subnet query: %w", err)
+	}
+
+	output := []providerSubnetCIDR{}
+	err = tx.Query(ctx, stmt, input).GetAll(&output)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		return nil, errors.Capture(err)
+	}
+
+	result := transform.SliceToMap(output, func(in providerSubnetCIDR) (string, providerSubnetCIDR) {
+		return in.ProviderID, in
+	})
+
+	return result, nil
+}
+
+// ensureAddressSubnetSuffix return an address including the subnet's suffix
+// and the subnetUUID. If the address is not an IPv4 nor IPv6 address, no
+// suffix is added.
+func ensureAddressSubnetSuffix(value, subnetProviderID string, data map[string]providerSubnetCIDR) (string, string, error) {
+	// Getting the subnet is helpful, but not essential.
+	subnet, _ := data[subnetProviderID]
+
+	parts := strings.Split(value, "/")
+	if len(parts) == 2 {
+		// Best case, the address has a CIDR suffix.
+		return value, subnet.SubnetUUID, nil
+	}
+
+	// Find the CIDR suffix, try the subnetCIDR first, if
+	// not define based on address type.
+	var suffix string
+	subnetCIDRParts := strings.Split(subnet.CIDR, "/")
+	switch len(subnetCIDRParts) {
+	case 2:
+		suffix = subnetCIDRParts[1]
+	case 1:
+		addType := corenetwork.DeriveAddressType(value)
```

Comment:
If the subnet has no CIDR, that means either the subnet is not existing/subnetProviderID is not set or that we have an address that isn't in that subnet's range. In both cases, lets return empty string for subnet UUID in addition to suffixing with the /32 or /128.

Reply:
https://warthogs.atlassian.net/browse/JUJU-8313